replace ROSThreadObj with std::thread

sarimabbas opened the pull request:
some cleanup required. hopefully if travis succeeds i will remove the extra print messages and comments

sarimabbas committed:
replaced ROSThreadObj with std::thread

alecive commented on baxter_collaboration_lib/include/robot_interface/robot_interface.h:
```diff
@@ -93,7 +95,8 @@ class RobotInterface
 
     // Mutex to protect joint state variable
-    pthread_mutex_t _mutex_jnts;
+    //pthread_mutex_t _mutex_jnts;
```
clean up all the things you commented out

alecive commented on baxter_collaboration_lib/include/robot_interface/robot_interface.h:
```diff
@@ -93,7 +95,8 @@ class RobotInterface
     // Mutex to protect joint state variable
-    pthread_mutex_t _mutex_jnts;
+    //pthread_mutex_t _mutex_jnts;
+    std::mutex _mtx_jnts;
```
call this variable `mutex_jnts` (or `mtx_jnts`), without the initial underscore

alecive commented on baxter_collaboration_lib/include/robot_interface/robot_interface.h:
```diff
@@ -116,7 +119,12 @@ class RobotInterface
+    //ROSThreadObj ctrl_thread;   // Internal thread that implements the controller server
+    std::thread ctrl_thread;
+
+    bool _thread_kill = false;
```
remove initial underscore, and call this `ctrl_thread_close_flag`. Plus, you should not assign values to variable in the header file, but in the constructor!

alecive commented on baxter_collaboration_lib/include/robot_interface/robot_interface.h:
```diff
@@ -116,7 +119,12 @@ class RobotInterface
+    std::thread ctrl_thread;
+
+    bool _thread_kill = false;
+    std::mutex _mtx_thread_kill;
```
`mtx_ctrl_thread_close_flag`

alecive commented on baxter_collaboration_lib/include/robot_interface/robot_interface.h:
```diff
@@ -254,7 +263,7 @@ class RobotInterface
      * @return  true/false if success failure (NOT in the POSIX way)
      */
-    bool closeThread();
+    //bool closeThread();
```
everything that is not needed, just remove it instead of commenting it out.

alecive commented on baxter_collaboration_lib/src/robot_interface/robot_interface.cpp:
```diff
@@ -103,26 +103,42 @@ RobotInterface::RobotInterface(string _name, string _limb, bool _use_robot, doub
     ros::Rate r(ctrl_freq);
 
-    while (RobotInterface::ok())
+    while ((not ros::isShuttingDown()) && (getThreadKill() == false))
```
maybe now is better to move `not ros::isShuttingDown()` to `ros::ok()` (test please)

alecive commented on baxter_collaboration_lib/src/robot_interface/robot_interface.cpp:
```diff
@@ -207,10 +223,12 @@ void RobotInterface::ThreadEntry()
         }
+
         r.sleep();
+        ROS_INFO("[%s] This is the value of r: %g", getLimb().c_str(), r.expectedCycleTime().toSec());
```
comment this out

alecive commented on baxter_collaboration_lib/src/robot_interface/robot_interface.cpp:
```diff
@@ -409,9 +427,11 @@ void RobotInterface::setCtrlRunning(bool _flag)
+    _mtx_ctrl.lock();
     is_ctrl_running = _flag;
-    pthread_mutex_unlock(&_mutex_ctrl);
+    _mtx_ctrl.unlock();
```
maybe here you can use a lock guard as well?

alecive commented on baxter_collaboration_lib/src/robot_interface/robot_interface.cpp:
```diff
@@ -423,9 +443,11 @@ bool RobotInterface::isCtrlRunning()
+    _mtx_ctrl.lock();
     res = is_ctrl_running;
-    pthread_mutex_unlock(&_mutex_ctrl);
+    _mtx_ctrl.unlock();
```
lock guard also here?

alecive commented:
Okay, I just pointed out some minor cleanup issues. Mainly:

 * remove commented out parts of the code
 * use `lock_guards` everywhere

sarimabbas committed:
cleanup

alecive commented on baxter_collaboration_lib/include/robot_interface/robot_interface.h:
```diff
@@ -140,7 +145,7 @@ class RobotInterface
 
     // Mutex to protect the control flag
-    pthread_mutex_t _mutex_ctrl;
+    std::mutex _mtx_ctrl;
```
remove first underscore

alecive commented on baxter_collaboration_lib/include/robot_interface/robot_interface.h:
```diff
@@ -651,6 +637,12 @@ class RobotInterface
+     * Safely manipulate the boolean needed to kill the thread entry
+     */
+    void setThreadKill(bool arg);
+    bool getThreadKill();
```
these function names should be aligned with the change in the variable name

alecive commented on baxter_collaboration_lib/src/robot_interface/robot_interface.cpp:
```diff
@@ -14,16 +14,10 @@ RobotInterface::RobotInterface(string _name, string _limb, bool _use_robot, doub
                                ik_solver(_limb, _use_robot), use_trac_ik(_use_trac_ik), ctrl_freq(_ctrl_freq),
                                filt_force{0.0, 0.0, 0.0}, filt_change{0.0, 0.0, 0.0}, time_filt_last_updated(ros::Time::now()),
-                               is_coll_av_on(false), is_coll_det_on(false), use_cart_ctrl(_use_cart_ctrl), is_ctrl_running(false),
+                               is_coll_av_on(false), is_coll_det_on(false), ctrl_thread_close_flag(false), use_cart_ctrl(_use_cart_ctrl), is_ctrl_running(false),
```
Good! :+1:

alecive commented on baxter_collaboration_lib/src/robot_interface/robot_interface.cpp:
```diff
@@ -103,26 +97,27 @@ RobotInterface::RobotInterface(string _name, string _limb, bool _use_robot, doub
 {
-    return ctrl_thread.start(ThreadEntryFunc, this);
+    ctrl_thread = std::thread(ThreadEntryFunc, this);
+	return ctrl_thread.joinable(); // std::thread::joinable checks if thread identifies as active thread of execution
```
how's the indentation here?

sarimabbas commented:
* ros::ok() seems to work
* put lockguards everywhere except `bool RobotInterface::computeIK(double px, double py, double pz...)` because the original lock/unlock was used to protect one line of code and I was cautious of blocking the rest of the code until the guard went out of scope.
* changed the variable names
* removed stray comments (but did not remove the debugging comments which predated my work)

alecive commented on baxter_collaboration_lib/src/robot_interface/robot_interface.cpp:
```diff
@@ -103,26 +97,27 @@ RobotInterface::RobotInterface(string _name, string _limb, bool _use_robot, doub
     ros::Rate r(ctrl_freq);
 
-    while (RobotInterface::ok())
+    while ((ros::ok()) && (getThreadKill() == false))
```
move this to  `(ros::ok() && not getThreadKill())`

alecive commented on baxter_collaboration_lib/src/robot_interface/robot_interface.cpp:
```diff
@@ -704,9 +694,9 @@ bool RobotInterface::computeIK(double px, double py, double pz,
+        mutex_jnts.lock();
         ik_srv.request.seed_angles.push_back(curr_jnts);
-        pthread_mutex_unlock(&_mutex_jnts);
+        mutex_jnts.unlock();
```
`lock_guard` here?

sarimabbas committed:
more cleanup

sarimabbas committed:
Revert "more cleanup"

This reverts commit ab2d30b86a13be5d1666b4ba478ba6fdab74e565.

sarimabbas committed:
better cleanup

alecive commented:
Ok good, now move onto the robot :+1:

alecive committed:
[robot_interface] Minor fixes to how thread is startedd.

alecive committed:
Minor fix to how mutexes are handled. Ready to merge
